packages/realm/src/app-services/BaseSubscriptionSet.ts

```diff
@@ -0,0 +1,216 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { MutableSubscriptionSet, Realm, Subscription, SubscriptionSet, assert, binding } from "../internal";
+
+/**
+ * Enum representing the state of a {@link SubscriptionSet}.
+ */
+export enum SubscriptionsState {
+  /**
+   * The subscription update has been persisted locally, but the server hasn't
+   * yet returned all the data that matched the updated subscription queries.
+   */
+  Pending = "pending",
+
+  /**
+   * The server has acknowledged the subscription and sent all the data that
+   * matched the subscription queries at the time the SubscriptionSet was
+   * updated. The server is now in steady-state synchronization mode where it
+   * will stream updates as they come.
+   */
+  Complete = "complete",
+
+  /**
+   * The server has returned an error and synchronization is paused for this
+   * Realm. To view the actual error, use `Subscriptions.error`.
+   *
+   * You can still use {@link SubscriptionSet.update} to update the subscriptions,
+   * and if the new update doesn't trigger an error, synchronization will be restarted.
+   */
+  Error = "error",
+
+  /**
+   * The SubscriptionSet has been superseded by an updated one. This typically means
+   * that someone has called {@link SubscriptionSet.update} on a different instance
+   * of the `Subscriptions`. You should not use a superseded SubscriptionSet,
+   * and instead obtain a new instance from {@link Realm.subscriptions}.
+   */
+  Superseded = "superseded",
+}
+
+const DEFAULT_PROPERTY_DESCRIPTOR: PropertyDescriptor = { configurable: true, enumerable: true, writable: true };
+const PROXY_HANDLER: ProxyHandler<BaseSubscriptionSet> = {
+  get(target, prop) {
+    if (Reflect.has(target, prop)) {
+      return Reflect.get(target, prop);
+    }
+    if (typeof prop === "string") {
+      const BASE = 10;
+      const index = Number.parseInt(prop, BASE);
+      // TODO: Consider catching an error from access out of bounds, instead of checking the length, to optimize for the hot path
+      if (!Number.isNaN(index) && index >= 0 && index < target.length) {
+        return target.get(index);
+      }
+    }
+  },
+  ownKeys(target) {
+    return Reflect.ownKeys(target).concat([...target.keys()].map(String));
+  },
+  getOwnPropertyDescriptor(target, prop) {
+    if (Reflect.has(target, prop)) {
+      return Reflect.getOwnPropertyDescriptor(target, prop);
+    } else if (typeof prop === "string") {
+      const BASE = 10;
+      const index = Number.parseInt(prop, BASE);
+      if (index < target.length) {
+        return DEFAULT_PROPERTY_DESCRIPTOR;
+      }
+    }
```

Comment:
What will be returned if the conditions are not met? From [`Reflect.ownKeys()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect/ownKeys) my guess is that `[]` should be returned.

Reply:
`getOwnPropertyDescriptor()` will return `undefined`. To make it more similar to JS arrays, we should probably also return `undefined` for negative indices. So rewriting the condition to `index >= 0 && index < target.length`.

Reply:
Yes, please return `undefined` for negative indices.